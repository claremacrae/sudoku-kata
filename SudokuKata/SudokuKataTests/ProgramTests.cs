using ApprovalTests;
using SudokuKata;
using System;
using ApprovalTests.Namers;
using SudokuKata.Utilities;
using SudokuKataTests.TestHelpers;
using Xunit;
using Xunit.Abstractions;

namespace SudokuKataTests
{
    [Collection("Sequential")]
    //  ^^^ Prevent running in parallel with other tests that write to console
    //  This test works by capturing any output to standard output.
    //  It must not run in parallel with any other tests that also write to standard output.
    //  Any other tests that also capture console output should also be put in the same Collection.
    //  See https://stackoverflow.com/a/47174842/104370
    public class ProgramTests
    {
        private readonly ITestOutputHelper _testOutputHelper;

        public ProgramTests(ITestOutputHelper testOutputHelper)
        {
            _testOutputHelper = testOutputHelper;
        }

        [Fact]
        public void TestSeed()
        {
            var errors = 0;
            for (int seed = 0; seed < 20; seed++)
            {
                var section = $"{seed:D4}";
                using (var cleanup = NamerFactory.AsEnvironmentSpecificTest("" + section))
                {
                    try
                    {
                        VerifySudokuForSeed(seed);
                    }
                    catch (Exception e)
                    {
                        _testOutputHelper.WriteLine(e.ToString());
                        errors += 1;
                    }
                }
            }
            Assert.Equal(0, errors);
        }
        
        private static void VerifySudokuForSeed(int seed)
        {
            var currentConsoleOut = Console.Out;
            using (var consoleOutput = new ConsoleUtilities.ConsoleOutput())
            {
                Program.Play(new Random(seed));
                string s = consoleOutput.GetOuput();
                Approvals.Verify(s);
            }
            Assert.Equal(currentConsoleOut, Console.Out);
        }
        
        [Fact]
        public void ShouldSolveGrid()
        {
            // See Easy https://www.websudoku.com/?level=1&set_id=2417252423
            var input    = "004509000500100048090020560000030016340010025720060000059040070270001003000702600";
            var expected = "164589237532176948897423561985234716346917825721865394659348172278691453413752689";

            var inputDigits = new int[81];
            Assert.Equal(81, input.Length);
            for (var i = 0; i < input.Length; i++)
            {
                var character = input[i];
                var digit = (int)char.GetNumericValue(character);
                inputDigits[i] = digit;
            }

            var finalState = new int[81];
            var grid = new Grid(inputDigits, finalState);
            Program.SolveBoard(new Random(0), grid);
            Assert.Equal(expected, GridStringifier.ConvertToCode(grid._state));
        }
    }
}