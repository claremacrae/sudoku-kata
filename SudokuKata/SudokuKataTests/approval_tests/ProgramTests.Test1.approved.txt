
Final look of the solved board:
+---+---+---+
|985|163|274|
|362|749|581|
|741|582|963|
+---+---+---+
|137|628|459|
|456|391|827|
|298|475|316|
+---+---+---+
|619|254|738|
|574|836|192|
|823|917|645|
+---+---+---+

Starting look of the board to solve:
+---+---+---+
|.8.|.6.|274|
|3..|...|...|
|7..|.82|...|
+---+---+---+
|..7|.28|459|
|..6|...|...|
|..8|.7.|3..|
+---+---+---+
|...|...|.3.|
|...|836|19.|
|823|...|64.|
+---+---+---+

================================================================================

(4, 1) can only contain 1.
+---+---+---+
|.8.|.6.|274|
|3..|...|...|
|7..|.82|...|
+---+---+---+
|1.7|.28|459|
|..6|...|...|
|..8|.7.|3..|
+---+---+---+
|...|...|.3.|
|...|836|19.|
|823|...|64.|
+---+---+---+
Code: 080060274300000000700082000107028459006000000008070300000000030000836190823000640

(4, 2) can only contain 3.
+---+---+---+
|.8.|.6.|274|
|3..|...|...|
|7..|.82|...|
+---+---+---+
|137|.28|459|
|..6|...|...|
|..8|.7.|3..|
+---+---+---+
|...|...|.3.|
|...|836|19.|
|823|...|64.|
+---+---+---+
Code: 080060274300000000700082000137028459006000000008070300000000030000836190823000640

(4, 4) can only contain 6.
+---+---+---+
|.8.|.6.|274|
|3..|...|...|
|7..|.82|...|
+---+---+---+
|137|628|459|
|..6|...|...|
|..8|.7.|3..|
+---+---+---+
|...|...|.3.|
|...|836|19.|
|823|...|64.|
+---+---+---+
Code: 080060274300000000700082000137628459006000000008070300000000030000836190823000640

Column #4 can contain 2 only at (7, 4).
+---+---+---+
|.8.|.6.|274|
|3..|...|...|
|7..|.82|...|
+---+---+---+
|137|628|459|
|..6|...|...|
|..8|.7.|3..|
+---+---+---+
|...|2..|.3.|
|...|836|19.|
|823|...|64.|
+---+---+---+
Code: 080060274300000000700082000137628459006000000008070300000200030000836190823000640

Block (1, 1) can contain 2 only at (2, 3).
+---+---+---+
|.8.|.6.|274|
|3.2|...|...|
|7..|.82|...|
+---+---+---+
|137|628|459|
|..6|...|...|
|..8|.7.|3..|
+---+---+---+
|...|2..|.3.|
|...|836|19.|
|823|...|64.|
+---+---+---+
Code: 080060274302000000700082000137628459006000000008070300000200030000836190823000640

Column #1 can contain 6 only at (7, 1).
+---+---+---+
|.8.|.6.|274|
|3.2|...|...|
|7..|.82|...|
+---+---+---+
|137|628|459|
|..6|...|...|
|..8|.7.|3..|
+---+---+---+
|6..|2..|.3.|
|...|836|19.|
|823|...|64.|
+---+---+---+
Code: 080060274302000000700082000137628459006000000008070300600200030000836190823000640

Row #8 can contain 2 only at (8, 9).
+---+---+---+
|.8.|.6.|274|
|3.2|...|...|
|7..|.82|...|
+---+---+---+
|137|628|459|
|..6|...|...|
|..8|.7.|3..|
+---+---+---+
|6..|2..|.3.|
|...|836|192|
|823|...|64.|
+---+---+---+
Code: 080060274302000000700082000137628459006000000008070300600200030000836192823000640

Column #9 can contain 3 only at (3, 9).
+---+---+---+
|.8.|.6.|274|
|3.2|...|...|
|7..|.82|..3|
+---+---+---+
|137|628|459|
|..6|...|...|
|..8|.7.|3..|
+---+---+---+
|6..|2..|.3.|
|...|836|192|
|823|...|64.|
+---+---+---+
Code: 080060274302000000700082003137628459006000000008070300600200030000836192823000640

Row #8 can contain 7 only at (8, 2).
+---+---+---+
|.8.|.6.|274|
|3.2|...|...|
|7..|.82|..3|
+---+---+---+
|137|628|459|
|..6|...|...|
|..8|.7.|3..|
+---+---+---+
|6..|2..|.3.|
|.7.|836|192|
|823|...|64.|
+---+---+---+
Code: 080060274302000000700082003137628459006000000008070300600200030070836192823000640

Values 4 and 5 in block (3, 1) are in cells (8, 1) and (8, 3).
4, 5 cannot appear in (7, 2).
4, 5 cannot appear in (7, 3).
Values 1 and 9 in row #7 are in cells (7, 2) and (7, 3).
1, 9 cannot appear in (7, 5).
1, 9 cannot appear in (7, 6).
Guessing that 1 and 9 are arbitrary in row #7 (multiple solutions): Pick 1->(7, 2), 9->(7, 3).
+---+---+---+
|.8.|.6.|274|
|3.2|...|...|
|7..|.82|..3|
+---+---+---+
|137|628|459|
|..6|...|...|
|..8|.7.|3..|
+---+---+---+
|619|2..|.3.|
|.7.|836|192|
|823|...|64.|
+---+---+---+
Code: 080060274302000000700082003137628459006000000008070300619200030070836192823000640

Guessing that 4 and 5 are arbitrary in row #8 (multiple solutions): Pick 5->(8, 1), 4->(8, 3).
+---+---+---+
|.8.|.6.|274|
|3.2|...|...|
|7..|.82|..3|
+---+---+---+
|137|628|459|
|..6|...|...|
|..8|.7.|3..|
+---+---+---+
|619|2..|.3.|
|574|836|192|
|823|...|64.|
+---+---+---+
Code: 080060274302000000700082003137628459006000000008070300619200030574836192823000640

(1, 1) can only contain 9.
+---+---+---+
|98.|.6.|274|
|3.2|...|...|
|7..|.82|..3|
+---+---+---+
|137|628|459|
|..6|...|...|
|..8|.7.|3..|
+---+---+---+
|619|2..|.3.|
|574|836|192|
|823|...|64.|
+---+---+---+
Code: 980060274302000000700082003137628459006000000008070300619200030574836192823000640

Values 1 and 5 in block (1, 1) are in cells (1, 3) and (3, 3).
5 cannot appear in (2, 2).
5 cannot appear in (3, 2).
Values 2 and 4 in block (2, 1) are in cells (5, 1) and (6, 1).
4 cannot appear in (5, 2).
4 cannot appear in (6, 2).
Guessing that 1 and 5 are arbitrary in column #3 (multiple solutions): Pick 5->(1, 3), 1->(3, 3).
+---+---+---+
|985|.6.|274|
|3.2|...|...|
|7.1|.82|..3|
+---+---+---+
|137|628|459|
|..6|...|...|
|..8|.7.|3..|
+---+---+---+
|619|2..|.3.|
|574|836|192|
|823|...|64.|
+---+---+---+
Code: 985060274302000000701082003137628459006000000008070300619200030574836192823000640

(3, 8) can only contain 6.
+---+---+---+
|985|.6.|274|
|3.2|...|...|
|7.1|.82|.63|
+---+---+---+
|137|628|459|
|..6|...|...|
|..8|.7.|3..|
+---+---+---+
|619|2..|.3.|
|574|836|192|
|823|...|64.|
+---+---+---+
Code: 985060274302000000701082063137628459006000000008070300619200030574836192823000640

(3, 2) can only contain 4.
+---+---+---+
|985|.6.|274|
|3.2|...|...|
|741|.82|.63|
+---+---+---+
|137|628|459|
|..6|...|...|
|..8|.7.|3..|
+---+---+---+
|619|2..|.3.|
|574|836|192|
|823|...|64.|
+---+---+---+
Code: 985060274302000000741082063137628459006000000008070300619200030574836192823000640

(2, 2) can only contain 6.
+---+---+---+
|985|.6.|274|
|362|...|...|
|741|.82|.63|
+---+---+---+
|137|628|459|
|..6|...|...|
|..8|.7.|3..|
+---+---+---+
|619|2..|.3.|
|574|836|192|
|823|...|64.|
+---+---+---+
Code: 985060274362000000741082063137628459006000000008070300619200030574836192823000640

Block (2, 3) can contain 6 only at (6, 9).
+---+---+---+
|985|.6.|274|
|362|...|...|
|741|.82|.63|
+---+---+---+
|137|628|459|
|..6|...|...|
|..8|.7.|3.6|
+---+---+---+
|619|2..|.3.|
|574|836|192|
|823|...|64.|
+---+---+---+
Code: 985060274362000000741082063137628459006000000008070306619200030574836192823000640

Values 1 and 3 in block (1, 2) are in cells (1, 4) and (1, 6).
1 cannot appear in (2, 4).
1 cannot appear in (2, 5).
1 cannot appear in (2, 6).
Guessing that 1 and 3 are arbitrary in row #1 (multiple solutions): Pick 1->(1, 4), 3->(1, 6).
+---+---+---+
|985|163|274|
|362|...|...|
|741|.82|.63|
+---+---+---+
|137|628|459|
|..6|...|...|
|..8|.7.|3.6|
+---+---+---+
|619|2..|.3.|
|574|836|192|
|823|...|64.|
+---+---+---+
Code: 985163274362000000741082063137628459006000000008070306619200030574836192823000640

Column #4 can contain 3 only at (5, 4).
+---+---+---+
|985|163|274|
|362|...|...|
|741|.82|.63|
+---+---+---+
|137|628|459|
|..6|3..|...|
|..8|.7.|3.6|
+---+---+---+
|619|2..|.3.|
|574|836|192|
|823|...|64.|
+---+---+---+
Code: 985163274362000000741082063137628459006300000008070306619200030574836192823000640

Guessing that 5 and 9 are arbitrary in row #3 (multiple solutions): Pick 5->(3, 4), 9->(3, 7).
+---+---+---+
|985|163|274|
|362|...|...|
|741|582|963|
+---+---+---+
|137|628|459|
|..6|3..|...|
|..8|.7.|3.6|
+---+---+---+
|619|2..|.3.|
|574|836|192|
|823|...|64.|
+---+---+---+
Code: 985163274362000000741582963137628459006300000008070306619200030574836192823000640

Guessing that 5 and 9 are arbitrary in column #2 (multiple solutions): Pick 5->(5, 2), 9->(6, 2).
+---+---+---+
|985|163|274|
|362|...|...|
|741|582|963|
+---+---+---+
|137|628|459|
|.56|3..|...|
|.98|.7.|3.6|
+---+---+---+
|619|2..|.3.|
|574|836|192|
|823|...|64.|
+---+---+---+
Code: 985163274362000000741582963137628459056300000098070306619200030574836192823000640

(6, 4) can only contain 4.
+---+---+---+
|985|163|274|
|362|...|...|
|741|582|963|
+---+---+---+
|137|628|459|
|.56|3..|...|
|.98|47.|3.6|
+---+---+---+
|619|2..|.3.|
|574|836|192|
|823|...|64.|
+---+---+---+
Code: 985163274362000000741582963137628459056300000098470306619200030574836192823000640

(6, 1) can only contain 2.
+---+---+---+
|985|163|274|
|362|...|...|
|741|582|963|
+---+---+---+
|137|628|459|
|.56|3..|...|
|298|47.|3.6|
+---+---+---+
|619|2..|.3.|
|574|836|192|
|823|...|64.|
+---+---+---+
Code: 985163274362000000741582963137628459056300000298470306619200030574836192823000640

(5, 1) can only contain 4.
+---+---+---+
|985|163|274|
|362|...|...|
|741|582|963|
+---+---+---+
|137|628|459|
|456|3..|...|
|298|47.|3.6|
+---+---+---+
|619|2..|.3.|
|574|836|192|
|823|...|64.|
+---+---+---+
Code: 985163274362000000741582963137628459456300000298470306619200030574836192823000640

(6, 8) can only contain 1.
+---+---+---+
|985|163|274|
|362|...|...|
|741|582|963|
+---+---+---+
|137|628|459|
|456|3..|...|
|298|47.|316|
+---+---+---+
|619|2..|.3.|
|574|836|192|
|823|...|64.|
+---+---+---+
Code: 985163274362000000741582963137628459456300000298470316619200030574836192823000640

(2, 8) can only contain 8.
+---+---+---+
|985|163|274|
|362|...|.8.|
|741|582|963|
+---+---+---+
|137|628|459|
|456|3..|...|
|298|47.|316|
+---+---+---+
|619|2..|.3.|
|574|836|192|
|823|...|64.|
+---+---+---+
Code: 985163274362000080741582963137628459456300000298470316619200030574836192823000640

(5, 8) can only contain 2.
+---+---+---+
|985|163|274|
|362|...|.8.|
|741|582|963|
+---+---+---+
|137|628|459|
|456|3..|.2.|
|298|47.|316|
+---+---+---+
|619|2..|.3.|
|574|836|192|
|823|...|64.|
+---+---+---+
Code: 985163274362000080741582963137628459456300020298470316619200030574836192823000640

(6, 6) can only contain 5.
+---+---+---+
|985|163|274|
|362|...|.8.|
|741|582|963|
+---+---+---+
|137|628|459|
|456|3..|.2.|
|298|475|316|
+---+---+---+
|619|2..|.3.|
|574|836|192|
|823|...|64.|
+---+---+---+
Code: 985163274362000080741582963137628459456300020298475316619200030574836192823000640

(2, 7) can only contain 5.
+---+---+---+
|985|163|274|
|362|...|58.|
|741|582|963|
+---+---+---+
|137|628|459|
|456|3..|.2.|
|298|475|316|
+---+---+---+
|619|2..|.3.|
|574|836|192|
|823|...|64.|
+---+---+---+
Code: 985163274362000580741582963137628459456300020298475316619200030574836192823000640

(2, 9) can only contain 1.
+---+---+---+
|985|163|274|
|362|...|581|
|741|582|963|
+---+---+---+
|137|628|459|
|456|3..|.2.|
|298|475|316|
+---+---+---+
|619|2..|.3.|
|574|836|192|
|823|...|64.|
+---+---+---+
Code: 985163274362000581741582963137628459456300020298475316619200030574836192823000640

Guessing that 7 and 9 are arbitrary in column #4 (multiple solutions): Pick 7->(2, 4), 9->(9, 4).
+---+---+---+
|985|163|274|
|362|7..|581|
|741|582|963|
+---+---+---+
|137|628|459|
|456|3..|.2.|
|298|475|316|
+---+---+---+
|619|2..|.3.|
|574|836|192|
|823|9..|64.|
+---+---+---+
Code: 985163274362700581741582963137628459456300020298475316619200030574836192823900640

Guessing that 7 and 8 are arbitrary in column #7 (multiple solutions): Pick 8->(5, 7), 7->(7, 7).
+---+---+---+
|985|163|274|
|362|7..|581|
|741|582|963|
+---+---+---+
|137|628|459|
|456|3..|82.|
|298|475|316|
+---+---+---+
|619|2..|73.|
|574|836|192|
|823|9..|64.|
+---+---+---+
Code: 985163274362700581741582963137628459456300820298475316619200730574836192823900640

(9, 9) can only contain 5.
+---+---+---+
|985|163|274|
|362|7..|581|
|741|582|963|
+---+---+---+
|137|628|459|
|456|3..|82.|
|298|475|316|
+---+---+---+
|619|2..|73.|
|574|836|192|
|823|9..|645|
+---+---+---+
Code: 985163274362700581741582963137628459456300820298475316619200730574836192823900645

(9, 5) can only contain 1.
+---+---+---+
|985|163|274|
|362|7..|581|
|741|582|963|
+---+---+---+
|137|628|459|
|456|3..|82.|
|298|475|316|
+---+---+---+
|619|2..|73.|
|574|836|192|
|823|91.|645|
+---+---+---+
Code: 985163274362700581741582963137628459456300820298475316619200730574836192823910645

(7, 9) can only contain 8.
+---+---+---+
|985|163|274|
|362|7..|581|
|741|582|963|
+---+---+---+
|137|628|459|
|456|3..|82.|
|298|475|316|
+---+---+---+
|619|2..|738|
|574|836|192|
|823|91.|645|
+---+---+---+
Code: 985163274362700581741582963137628459456300820298475316619200738574836192823910645

(7, 6) can only contain 4.
+---+---+---+
|985|163|274|
|362|7..|581|
|741|582|963|
+---+---+---+
|137|628|459|
|456|3..|82.|
|298|475|316|
+---+---+---+
|619|2.4|738|
|574|836|192|
|823|91.|645|
+---+---+---+
Code: 985163274362700581741582963137628459456300820298475316619204738574836192823910645

(5, 5) can only contain 9.
+---+---+---+
|985|163|274|
|362|7..|581|
|741|582|963|
+---+---+---+
|137|628|459|
|456|39.|82.|
|298|475|316|
+---+---+---+
|619|2.4|738|
|574|836|192|
|823|91.|645|
+---+---+---+
Code: 985163274362700581741582963137628459456390820298475316619204738574836192823910645

(5, 9) can only contain 7.
+---+---+---+
|985|163|274|
|362|7..|581|
|741|582|963|
+---+---+---+
|137|628|459|
|456|39.|827|
|298|475|316|
+---+---+---+
|619|2.4|738|
|574|836|192|
|823|91.|645|
+---+---+---+
Code: 985163274362700581741582963137628459456390827298475316619204738574836192823910645

(2, 5) can only contain 4.
+---+---+---+
|985|163|274|
|362|74.|581|
|741|582|963|
+---+---+---+
|137|628|459|
|456|39.|827|
|298|475|316|
+---+---+---+
|619|2.4|738|
|574|836|192|
|823|91.|645|
+---+---+---+
Code: 985163274362740581741582963137628459456390827298475316619204738574836192823910645

(9, 6) can only contain 7.
+---+---+---+
|985|163|274|
|362|74.|581|
|741|582|963|
+---+---+---+
|137|628|459|
|456|39.|827|
|298|475|316|
+---+---+---+
|619|2.4|738|
|574|836|192|
|823|917|645|
+---+---+---+
Code: 985163274362740581741582963137628459456390827298475316619204738574836192823917645

(5, 6) can only contain 1.
+---+---+---+
|985|163|274|
|362|74.|581|
|741|582|963|
+---+---+---+
|137|628|459|
|456|391|827|
|298|475|316|
+---+---+---+
|619|2.4|738|
|574|836|192|
|823|917|645|
+---+---+---+
Code: 985163274362740581741582963137628459456391827298475316619204738574836192823917645

(7, 5) can only contain 5.
+---+---+---+
|985|163|274|
|362|74.|581|
|741|582|963|
+---+---+---+
|137|628|459|
|456|391|827|
|298|475|316|
+---+---+---+
|619|254|738|
|574|836|192|
|823|917|645|
+---+---+---+
Code: 985163274362740581741582963137628459456391827298475316619254738574836192823917645

(2, 6) can only contain 9.
+---+---+---+
|985|163|274|
|362|749|581|
|741|582|963|
+---+---+---+
|137|628|459|
|456|391|827|
|298|475|316|
+---+---+---+
|619|254|738|
|574|836|192|
|823|917|645|
+---+---+---+
Code: 985163274362749581741582963137628459456391827298475316619254738574836192823917645

